pkg/awsutils/awsutils.go

```diff
@@ -458,42 +449,17 @@ func (cache *EC2InstanceMetadataCache) getENIMetadata(macStr string) (ENIMetadat
 	if err != nil {
 		return ENIMetadata{}, errors.Wrapf(err, "get ENI metadata: failed to retrieve IPs and CIDR for ENI: %s", eniMAC)
 	}
-	privateIPv4s, tags, _, err := cache.DescribeENI(eni)
-	if err != nil {
-		return ENIMetadata{}, errors.Wrapf(err, "get ENI metadata: failed to describe ENI: %s, %v", eniMAC, err)
-	}
-	// getIPsAndCIDR() queries IMDS for IPv4 addresses attached to the ENI.
-	// DescribeENI() calls the DescribeNetworkInterfaces AWS API call, which
-	// technically should be the source of truth and contain the freshest
-	// information. Let's just do a quick scan here and output some diagnostic
-	// messages if we find stale info in the IMDS result.
-	imdsIPv4Set := sets.NewString(imdsIPv4s...)
-	privateIPv4Set := sets.String{}
-	for _, privateIPv4 := range privateIPv4s {
-		privateIPv4Set.Insert(aws.StringValue(privateIPv4.PrivateIpAddress))
-	}
-	missingIMDS := privateIPv4Set.Difference(imdsIPv4Set).List()
-	missingDNI := imdsIPv4Set.Difference(privateIPv4Set).List()
-	if len(missingIMDS) > 0 {
-		strMissing := strings.Join(missingIMDS, ",")
-		log.Debugf("getENIMetadata: DescribeNetworkInterfaces(%s) yielded private IPv4 addresses %s that were not yet found in IMDS.", eni, strMissing)
-	}
-	if len(missingDNI) > 0 {
-		strMissing := strings.Join(missingDNI, ",")
-		log.Debugf("getENIMetadata: IMDS query yielded stale IPv4 addresses %s that were not found in DescribeNetworkInterfaces(%s).", strMissing, eni)
-	}
```

Comment:
The above code existed in order to determine that IMDS had stale information. `cache.DescribeENI()` is the only thing that actually ends up calling the DescribeNetworkInterfaces EC2 API call. By removing the call to EC2's DescribeNetworkInterfaces API call here, the caller of the cache `getENIMetadata()` call will now only look at potentially stale IMDS information. The only caller of `getENIMetadata()` happens to be the cache's `GetAttachedENIs()` method, which itself is called during ipamd's `nodeInit()` method:

https://github.com/aws/amazon-vpc-cni-k8s/blob/599fed32fdc8466b74e9468901a7d1bf4c228b42/pkg/ipamd/ipamd.go#L277

That method, unfortunately, MUST have the most up-to-date information returned to it, which means that DescribeNetworkInterfaces EC2 API call should be made, not IMDS. This is unfortunately what I was dealing with when I wrote the above code.

The other place that `GetAttachedENIs()` is called is in ipamd's `waitENIAttached()` method:

https://github.com/aws/amazon-vpc-cni-k8s/blob/599fed32fdc8466b74e9468901a7d1bf4c228b42/pkg/ipamd/ipamd.go#L799

In *this* case, we actually DO want to only check IMDS and not the EC2 DescribeNetworkInterfaces API call. So, I think the appropriate thing to do is to break up `GetAttachedENIs()` into an EC2-calling method and an IMDS-calling method and call one or the other depending on whether it is in `nodeInit()` or `waitENIAttached()`.

Otherwise, we really don't know whether the node is being initialized with stale data or not.

Thoughts?

Reply:
Great points @jaypipes. I know the reason for that diff and I did test a version where I had a "callEC2" bool, to disable it for the `waitENIAttached()` function. That said, I think there are two improvements to be made here. One is to move the tag check out from metadata, as has been done already, since that data is not available from IMDS. In addition we should make a separate `getENIInfoFromEC2()` function that does a batch-describe call of all attached ENIs to EC2 instead of calling them one by one, to be used by the init function to verify the status of all ENIs when the ipamd pod starts up. Will update the PR.